File: jaeseuk/kokotlin/src/main/kotlin/programmers/level1/FailureRate.kt
package programmers.level1

class FailureRate {
    fun solution(N: Int, stages: IntArray): IntArray {
        val players = IntArray(N + 1)
        var answer = arrayOf<Pair<Int, Double>>()
        var left = stages.size.toDouble()

        for (stage in stages) {
            players[stage - 1] += 1
        }

        for (i in 0 until N) {
            if (left != 0.0) {
                answer += i + 1 to players[i] / left
            } else {
                answer += i + 1 to 0.0
            }
            left -= players[i]
        }

        return answer
            .sortedWith(compareBy({-it.second}, {it.first}))
            .map { it.first }
            .toIntArray()
    }
}

File: jaeseuk/pypython/Programmers/level1/failure_rate.py
def solution(N, stages):
    fail = [0] * (N + 1)
    answer = []

    for stage in stages:
        fail[stage - 1] += 1

    for i in range(len(fail) - 1):
        players = sum(fail[i:])

        if players > 0:
            answer.append((i + 1, fail[i] / players))
        else:
            answer.append((i + 1, 0))

    answer.sort(key=lambda x: (-x[1], x[0]))

    return list(map(lambda x: x[0], answer))
